logger.go

```diff
@@ -14,147 +14,157 @@ type Logger struct {
 }
 
 // NewLogger creates a Logger instance with no additional data.
-func NewLogger() Logger {
-	return Logger{
+func NewLogger() *Logger {
+	return &Logger{
 		loggingT: &logging,
 	}
 }
 
 // WithContext creates a logger from a context.Context
-func WithContext(ctx context.Context) Logger {
+func WithContext(ctx context.Context) *Logger {
 	data, _ := ctx.Value(contextKeyData).([]interface{})
 	prefix, _ := ctx.Value(contextKeyPrefix).(string)
-	return Logger{
+	return &Logger{
 		loggingT: &logging,
 		data:     data,
 		prefix:   prefix,
 	}
 }
 
 // WithPrefix creates a Logger with a given prefix.
-func WithPrefix(prefix string) Logger {
-	return Logger{
+func WithPrefix(prefix string) *Logger {
+	return &Logger{
 		loggingT: &logging,
 		prefix:   prefix,
 	}
 }
 
 // WithData creates a Logger with a given set of backend data.
 // All args provided will be wrapped with glog.Data prior to being sent.
-func WithData(args ...interface{}) Logger {
-	return Logger{
+func WithData(args ...interface{}) *Logger {
+	return &Logger{
 		loggingT: &logging,
 		data:     args,
 	}
 }
 
 // WithPrefix creates a Logger from an existing logger with a specified prefix.
 // Any prefix on the input Logger will be replaced.
-func (l Logger) WithPrefix(prefix string) Logger {
-	newLogger := l
-	newLogger.prefix = prefix
-	return newLogger
+func (l *Logger) WithPrefix(prefix string) *Logger {
+	if l == nil {
+		return nil
+	}
+	return &Logger{
+		loggingT: l.loggingT,
+		data:     l.data,
```

Comment:
I'm sorry, but I'm still missing something. The data here will be shared between all loggers. The same exact slice. Meaning that the following code:

logger = glog.WithPrefix("handling things")
for _, thing := range things {
   logger.WithData(thing).WithPrefix(thing.Name).Info("Handling that") 
}
logger.Info("Done")

will produce:

INFO: handling things: <THING_1-NAME>: Handling that: data: [ thing { <THING_1-FIELDS> } ]
...
INFO: handling things: <THING_N-NAME>: Handling that: data: [ thing { <THING_N-FIELDS> } ]
INFO: handling things: Done: data: [ thing { <THING_N-FIELDS> } ] 

And that's assuming we don't do append on data. If we do, it will  contain the whole list of things.

If I understand correctly, the reason you wanted to make a copy of the logger each time and not return the same instance is to be able to use it when going 'down the call stack', e.g. use it in loops like above. But that means that on each 'pop', the logger must remain exactly as it was before going the extra level, which means that the lower levels cannot mutate it.

In general, it seems to me that applying different copy types ('deep' for prefix, 'shallow' for data) in the same struct in seemingly similar methods is sub-optimal conceptually (WithPrefix vs. WithData).

What am I missing?

Reply:
The append function creates a completely new slice. Unless we added code to actively change the contents of specific indexes, it won't be a problem.

Reply:
Good point re append! I never thought of that in depth, but it's indeed a totally different slice. 
However! It made me revisit the whole 'how-slices-work' topic and here is what I found:
Append returns a brand new slice, but the underlying array is the same. Therefore, concurrent routines may be stepping on each other's toes.
Here is an example: https://play.golang.org/p/ZH_XAoXAtME
Note the printouts for 'go func' start to be wrong once the main loop kicks in.
(I had to create the data with an initial capacity as otherwise almost every append a new underlying array is allocated and it's very hard to surface the issue)
Here is the same thing but with deep copy of data: https://play.golang.org/p/sbf9WLTRFxR

Reply:
This is true, but we don't export these slices, so nobody will be manipulating them but this code, and the tests should identify such a problem.

Reply:
Ah, see your point. Will make that change.

Reply:
Scratch that. Your play link almost had me convinced, but I added this as an actual test for safety and couldn't reproduce it.

What happens in your example is that you initialize the capacity of the backing array ahead of time, so the first few times, it doesn't need to re-allocate. If I remove the make([]string...) line:
https://play.golang.org/p/GQdsRH-z1nK
It works fine.

See: https://tour.golang.org/moretypes/15
"If the backing array of s is too small to fit all the given values a bigger array will be allocated. The returned slice will point to the newly allocated array."

It doesn't save any actual time, but there's no sense explicitly doing the copy when the runtime will do it for us. If someone tries to be clever, the test I've added will break.